text/0000-read-buf.md

```diff
@@ -0,0 +1,624 @@
+
+```rust
+impl Read for MyReader {
+    fn read_buf(&mut self, buf: &mut ReadBuf<'_>) -> io::Result<()> {
+        // Get access to the unwritten part of the buffer, making sure it has been fully initialized. Since `ReadBuf`
+        // tracks the initialization state of the buffer, this is "free" after the first time it's called.
+        let unwritten: &mut [u8] = buf.initialize_unwritten();
+
+        // Fill the whole buffer with some nonsense.
+        for (i, byte) in unwritten.iter_mut().enumerate() {
+            *byte = i as u8;
+        }
+
+        // And indicate that we've written the whole thing.
+        let len = unwritten.len();
+        buf.add_written(len);
+
+        Ok(())
+    }
+}
+```
+
+An unsafe `Read` implementation:
+
+```rust
+impl Read for TcpStream {
+    fn read_buf(&mut self, buf: &mut ReadBuf<'_>) -> io::Result<()> {
+        unsafe {
+            // Get access to the unwritten part of the buffer, without initializing it. This method is unsafe; we are
+            // responsible for ensuing that we don't "de-initialize" portions of it that have previously been
+            // initialized.
+            let unwritten: &mut [MaybeUninit<u8>] = buf.unwritten_mut();
+
+            // We're just delegating to the libc read function, which returns an `isize`. The return value indicates
+            // an error if negative and the number of bytes read otherwise.
+            let nread = libc::read(self.fd, unwritten.as_mut_ptr().cast::<libc::c_void>(), unwritten.len());
+
+            if nread < 0 {
+                return Err(io::Error::last_os_error());
+            }
+
+            let nread = nread as usize;
+            // If the read succeeded, tell the buffer that the read-to portion has been initialized. This method is
+            // unsafe; we are responsible for ensuring that this portion of the buffer has actually been initialized.
+            buf.assume_initialized(nread);
+            // And indicate that we've written the bytes as well. Unlike `assume_initialized`, this method is safe,
+            // and asserts that the written portion of the buffer does not advance beyond the initialized portion of
+            // the buffer. If we didn't call `assume_initialized` above, this call could panic.
+            buf.add_written(nread);
+
+            Ok(())
+        }
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+```rust
+/// A wrapper around a byte buffer that is incrementally written to and initialized.
+///
+/// This type is a sort of "double cursor". It tracks three regions in the buffer: a region at the beginning of the
+/// buffer that has been logically written to, a region that has been initialized at some point but not yet logically
+/// written to, and a region at the end that is fully uninitialized. The written-to region is guaranteed to be a
+/// subset of the initialized region.
+pub struct ReadBuf<'a> {
```

Comment:
Maybe explicitly state that `written <= initialized <= buf.len()` is an invariant of this type?

Reply:
The "the written-to region is guaranteed to be a subset of the initialized region" bit is trying to say that. Maybe it should be clarified?

Reply:
I just feel like this is worth repeating in math/Rust notation -- kind of the ultimate clarification. ;)

Reply:
It took me a while to understand that, having the mathematical notation would make it easier to understand.